### PR TITLE
NEW PageHeader component

### LINKED ISSUE #6090: PageHeader - add a PageHeader component
Implement a PageHeader component. It should fit well within Page and PageContent.

Some early exploration was done here: https://github.com/grommet/hpe-design-system/issues/2517

depends on https://github.com/grommet/hpe-design-system/issues/2542

### PR DESCRIPTION
**Pull request by taysea**
#### What does this PR do?

This is a proposal for a PageHeader component based on design explorations that have been done.

The PageHeader is a specialized Header that takes a set of properties (actions, context, subtitle, and title) that are then laid out by PageHeader in a Grid under the hood. The theme defines default gridAreas for where these properties are rendered, but the caller can override these in a custom theme if they desire a different layout.

```
// API
actions?: JSX.Element;
children?: JSX.Element[];
context?: JSX.Element;
gridProps?: GridProps;
subtitle?: string | JSX.Element;
title?: string | JSX.Element;
```

```
// default gridArea

['context, 'context']
['title', 'actions']
['subtitle', 'actions']
```

Using Grid is beneficial for the layout because it:
1. Allows the caller to define where they want the PageHeader elements to render (see Custom story).
2. Allows caller to render custom elements (see Children story).
3. Allows the caller to define the min/max widths that they want various content regions to span. For example, not allowing the title/subtitle to shrink less than width of "medium". (see Simple story and shrink window down, see base.js)

To-do:
- [x] How should actions align if title is large? See image of them top-aligned (current implementation), but should they be centered with the title? bottom-aligned w the title?:
Screen Shot 2022-04-26 at 4 08 42 PM
         A: provided ability for caller to pass box props and specify their own alignment and justification as needed.
- [x] For responsive button groups, how might the caller receive information regarding a gridArea's dimensions?
         A: Talked through a few approaches here, but do not feel that the solution may be implemented as a broader helper component which PageHeader and others may make use of.
- [x] What might the `context` prop be renamed to for more clarity on its purpose?
         A: Renamed to `parent` as it is clean, succinct, and accurate to parent-child relationship. No guessing involved.

#### Where should the reviewer start?
PageHeader.js

https://deploy-preview-6071--sad-tereshkova-173d07.netlify.app/?path=/story/layout-pageheader-simple--simple&globals=theme:hpe

#### What testing has been done on this PR?

Tests need to be added.

#### How should this be manually tested?

With stories, assess across various breakpoints.

#### Do Jest tests follow these best practices?

TBD. Tests need to be added.

- [ ] `screen` is used for querying.
- [ ] The correct query is used. (Refer to [this list of queries](https://testing-library.com/docs/queries/about/#priority))
- [ ] `userEvent` is used in place of `fireEvent`.
- [ ] `asFragment()` is used for snapshot testing.

#### What are the relevant issues?
Related to https://github.com/grommet/hpe-design-system/issues/2517
Closes #6090 

#### Do the grommet docs need to be updated?
Yes.

#### Should this PR be mentioned in the release notes?
Yes.

#### Is this change backwards compatible or is it a breaking change?
Backwards compatible.

### CODE REVIEW

**Comment by britt6612** [comment]
I like the `gridAreas` I did notice on smaller screens since the content went over the width of the grid a horizontal scroll was then present.

Not really something for you to solve but if we have 3-4 actions buttons I'm wondering how we can better present them. Since we don't want the title and subtitle to get below a certain width how can we do better at showing the actions buttons

Screen Shot 2022-04-15 at 2 21 48 PM

File: src/js/components/PageHeader/stories/Simple.js
Comment (britt6612) [inline src/js/components/PageHeader/stories/Simple.js:60]: I know this naming can be changed in the theme just putting it out there if we can think of a better name for `context`.

Comment (jcfilben) [reply src/js/components/PageHeader/stories/Simple.js:60]: Agree on changing the name for this. We are using this section for navigation elements but the naming `context` doesn't give me any clue that this is the area for the up button or breadcrumbs.

File: src/js/components/PageHeader/PageHeader.js
Comment (jcfilben) [inline src/js/components/PageHeader/PageHeader.js:13]: To me it feels more intuitive to have the rest on the Header. But maybe we should allow for passing additional grid props through `gridProps` similar to how on DropButton you can pass any additional drop props through `dropProps`

**Comment by taysea** [comment]
> Not really something for you to solve but if we have 3-4 actions buttons I'm wondering how we can better present them. Since we don't want the title and subtitle to get below a certain width how can we do better at showing the actions buttons
> 
> Screen Shot 2022-04-15 at 2 21 48 PM

Totally agree we need to give guidance here from a DS perspective. I left this in (hoping people would find the bad spots in the layout) so we can discuss.

File: src/js/components/PageHeader/PageHeader.js
Comment (taysea) [reply src/js/components/PageHeader/PageHeader.js:13]: That could play nicely, I like that suggestion.

File: src/js/components/PageHeader/stories/Simple.js
Comment (taysea) [reply src/js/components/PageHeader/stories/Simple.js:60]: Good comments. Originally I had `contextualNav`. Would this be more clear?

Comment (ericsoderberghp) [reply src/js/components/PageHeader/stories/Simple.js:60]: Good questions. I think `context` allows it to contain elements that provide "context" without them having to be navigational. For instance, it might make sense for a page about your car to have a title containing your license plate with a context of "Car", without having it be "Cars" as a navigational link. I suspect this example is on the 20% side though. But, I'm just trying to contemplate it through a wider lens of what someone might want to do with grommet. I could be OK with `contextNav`, as long as the caller can pass anything they want for it.

**Comment by ericsoderberghp** [comment]
> > Not really something for you to solve but if we have 3-4 actions buttons I'm wondering how we can better present them. Since we don't want the title and subtitle to get below a certain width how can we do better at showing the actions buttons
> > Screen Shot 2022-04-15 at 2 21 48 PM
> 
> Totally agree we need to give guidance here from a DS perspective. I left this in (hoping people would find the bad spots in the layout) so we can discuss.

I am hoping we can leverage some behavior and perhaps some components for the responsive behavior of Tabs, PageHeader actions, List or Table toolbar controls, and similar collections of controls. I think they all have a potential desire to covert into something like a DropButton with controls moved to within the Drop.

File: src/js/components/PageHeader/PageHeader.js
Comment (halocline) [reply src/js/components/PageHeader/PageHeader.js]: @jcfilben's comment was the direction I had been thinking as well.

File: src/js/components/PageHeader/PageHeader.js
Comment (halocline) [inline src/js/components/PageHeader/PageHeader.js:34]: Reiterating that I agree that the `...rest` belongs here at the outermost component to allow for passing of `id`, other standard props and box props.

File: src/js/components/PageHeader/PageHeader.js
Comment (halocline) [inline src/js/components/PageHeader/PageHeader.js:31]: left a related comment, should we allow for areas to vary by breakpoint?
```suggestion
    const customAreas = getCustomAreas(theme, size);
```

File: src/js/components/PageHeader/index.d.ts
Comment (halocline) [inline src/js/components/PageHeader/index.d.ts:13]: Can this be removed? I don't see it being consumed anywhere.

Comment (taysea) [reply src/js/components/PageHeader/index.d.ts:13]: These are the prop types for what we pass onto all components as "genericProps"

File: src/js/components/PageHeader/PageHeader.js
Comment (taysea) [reply src/js/components/PageHeader/PageHeader.js]: Done.

File: src/js/components/PageHeader/PageHeader.js
Comment (taysea) [reply src/js/components/PageHeader/PageHeader.js]: Done.

File: src/js/components/PageHeader/PageHeader.js
Comment (taysea) [reply src/js/components/PageHeader/PageHeader.js]: Done.

File: src/js/themes/base.js
Comment (halocline) [inline src/js/themes/base.js:1270]: small makes sense for HPE theme, but thinking Grommet should use the default heading size.

```suggestion
```

File: src/js/components/PageHeader/PageHeader.js
Comment (halocline) [inline src/js/components/PageHeader/PageHeader.js:54]:
```suggestion
```

**Review by halocline — COMMENTED** [review]
Still thinking about "For responsive button groups, how might the caller receive information regarding a gridArea's dimensions?"

File: src/js/components/PageHeader/stories/CustomAreas.js
Comment (taysea) [reply src/js/components/PageHeader/stories/CustomAreas.js:52]: Yes, you're right.

File: src/js/components/PageHeader/stories/CustomAreas.js
Comment (taysea) [reply src/js/components/PageHeader/stories/CustomAreas.js:30]: Yes, you're right, will update.

File: src/js/components/PageHeader/stories/CustomAreas.js
Comment (taysea) [reply src/js/components/PageHeader/stories/CustomAreas.js:22]: Thanks, had that fix locally but forgot to push it up (coming in next push).

File: src/js/components/PageHeader/PageHeader.js
Comment (taysea) [reply src/js/components/PageHeader/PageHeader.js:70]: Will also add one for the context / parent region (name tbd.)

File: src/js/components/PageHeader/stories/CustomAreas.js
Comment (taysea) [reply src/js/components/PageHeader/stories/CustomAreas.js]: Done.

File: src/js/components/PageHeader/PageHeader.js
Comment (taysea) [reply src/js/components/PageHeader/PageHeader.js]: Done.

File: src/js/themes/base.js
Comment (taysea) [reply src/js/themes/base.js]: Done.

File: src/js/components/PageHeader/PageHeader.js
Comment (taysea) [reply src/js/components/PageHeader/PageHeader.js]: Done.

File: src/js/components/PageHeader/stories/CustomAreas.js
Comment (taysea) [reply src/js/components/PageHeader/stories/CustomAreas.js]: Done.

File: src/js/components/PageHeader/PageHeader.js
Comment (taysea) [reply src/js/components/PageHeader/PageHeader.js]: Done. Removed.

File: src/js/components/PageHeader/PageHeader.js
Comment (taysea) [reply src/js/components/PageHeader/PageHeader.js]: Done.

File: src/js/components/PageHeader/stories/Simple.js
Comment (ericsoderberghp) [inline src/js/components/PageHeader/stories/Simple.js:40]: Can we have the Simple story use just a single Button in the `actions`, since I think that's the predominant case?

File: src/js/themes/base.js
Comment (halocline) [inline src/js/themes/base.js:1281]: Still waffling back and forth if we should a) group these breakpoints under a single property, b) keep them at this level, or c) or restructure to read component.property[breakpoint]? 

Looking through the rest of the theme file, I think I am seeing "c" most often.

@ericsoderberghp any thoughts? Preferences?

File: src/js/components/PageHeader/stories/Simple.js
Comment (halocline) [reply src/js/components/PageHeader/stories/Simple.js]: done.

File: src/js/themes/base.js
Comment (ericsoderberghp) [reply src/js/themes/base.js:1281]: We have been a bit inconsistent about this in the theme in the past. Sometimes we rooted more in aspect, with state variations beneath the aspect, as with `heading.level.1.size` or `tab.border.hover`. Other times, we rooted more in state, as with `button.hover.primary.background`. Our current direction is to go with the later, rooting more in state than in aspect. So for this, we'd do "b".
Admittedly, our theme patterns are bit of a mess as theming has grown.

**Comment by ericsoderberghp** [comment]
looking good, seems like tests are the next thing to add?

**Comment by halocline** [comment]
> looking good, seems like tests are the next thing to add?

writing them now

**Review by britt6612 — APPROVED** [review]
Looks good cleaned up nice! 
I would just make sure we add docs once this gets merged in! I can help with that if needed

File: src/js/components/PageHeader/stories/MultipleActions.js
Comment (britt6612) [inline src/js/components/PageHeader/stories/MultipleActions.js:31]: one small thing sorry do we need a `xsmall` here or are we not worried about that size for the hpe theme version

**Comment by ericsoderberghp** [comment]
Looking good to me. Waiting for Jessica and Taylor to take a look before final approval.

File: src/js/components/PageHeader/stories/MultipleActions.js
Comment (taysea) [reply src/js/components/PageHeader/stories/MultipleActions.js:31]: Since grommet only has `small - large` breakpoints, I don't think we would include it here but for an example on DS site we would implement with `xsmall`

File: src/js/components/PageHeader/PageHeader.js
Comment (taysea) [inline src/js/components/PageHeader/PageHeader.js:29]: Nice.

**Review by taysea — COMMENTED** [review]
Approved! (can't officially since it's my own PR)

**Review by jcfilben — APPROVED** [review]
Looks good!

File: src/js/components/PageHeader/stories/MultipleActions.js
Comment (britt6612) [reply src/js/components/PageHeader/stories/MultipleActions.js:31]: thats fine I just saw when I switched the theme on storybook

**Review by ericsoderberghp — APPROVED** [review]

**Pull request merged**
